util/extsort/disk_sorter.go

```diff
@@ -15,37 +15,606 @@
+	}
+	if v.refs.Add(-1) == 0 {
+		delete(p.mu.readers, fileNum)
+		return errors.Trace(v.reader.Close())
+	}
+	return nil
+}
+
+type sstIter struct {
+	iter    sstable.Iterator
+	key     *sstable.InternalKey
+	value   []byte
+	onClose func() error
+}
+
+func (si *sstIter) Seek(key []byte) bool {
+	si.key, si.value = si.iter.SeekGE(key)
+	return si.key != nil
+}
+
+func (si *sstIter) First() bool {
+	si.key, si.value = si.iter.SeekGE(nil)
+	return si.key != nil
+}
+
+func (si *sstIter) Next() bool {
+	si.key, si.value = si.iter.Next()
+	return si.key != nil
+}
+
+func (si *sstIter) Last() bool {
+	si.key, si.value = si.iter.Last()
+	return si.key != nil
+}
+
+func (si *sstIter) Valid() bool {
+	return si.key != nil
+}
+
+func (si *sstIter) Error() error {
+	return errors.Trace(si.iter.Error())
+}
+
+func (si *sstIter) UnsafeKey() []byte {
+	return si.key.UserKey
+}
+
+func (si *sstIter) UnsafeValue() []byte {
+	return si.value
+}
+
+func (si *sstIter) Close() error {
+	err := si.iter.Close()
+	if si.onClose != nil {
+		err = goerrors.Join(err, si.onClose())
+	}
+	return errors.Trace(err)
+}
+
+type mergingIter struct {
+	heap          mergingIterHeap
+	orderedFiles  []*fileMetadata // ordered by start key
+	nextFileIndex int
+	openIter      openIterFunc
+	curKey        []byte // only used in Next() to avoid alloc
+	err           error
+}
+
+type mergingIterItem struct {
+	Iterator
+	index int // index in orderedFiles
+}
+
+type mergingIterHeap []mergingIterItem
+
+func (h mergingIterHeap) Len() int {
+	return len(h)
+}
+
+func (h mergingIterHeap) Swap(i, j int) {
+	h[i], h[j] = h[j], h[i]
+}
+
+func (h mergingIterHeap) Less(i, j int) bool {
+	return bytes.Compare(h[i].UnsafeKey(), h[j].UnsafeKey()) < 0
+}
+
+func (h *mergingIterHeap) Push(x interface{}) {
+	*h = append(*h, x.(mergingIterItem))
+}
+
+func (h *mergingIterHeap) Pop() interface{} {
+	old := *h
+	n := len(old)
+	x := old[n-1]
+	*h = old[:n-1]
+	return x
+}
+
+type openIterFunc func(file *fileMetadata) (Iterator, error)
+
+// newMergingIter returns an iterator that merges the given files.
+// orderedFiles must be ordered by start key, otherwise it panics.
+func newMergingIter(orderedFiles []*fileMetadata, openIter openIterFunc) *mergingIter {
+	if !slices.IsSortedFunc(orderedFiles, func(a, b *fileMetadata) bool {
+		return bytes.Compare(a.startKey, b.startKey) < 0
+	}) {
+		panic("newMergingIter: orderedFiles are not ordered by start key")
+	}
+	return &mergingIter{
+		orderedFiles: orderedFiles,
+		openIter:     openIter,
+	}
+}
+
+func (m *mergingIter) Seek(key []byte) bool {
+	m.err = nil
+	oldHeap := m.heap
+	m.heap = nil
+
+	var toClose []mergingIterItem
+	openedIters := make(map[int]struct{})
+	for i, item := range oldHeap {
+		file := m.orderedFiles[item.index]
+		if bytes.Compare(key, file.startKey) >= 0 &&
+			bytes.Compare(key, file.endKey) < 0 &&
+			item.Seek(key) {
+			openedIters[item.index] = struct{}{}
+			heap.Push(&m.heap, item)
```

Comment:
heap.Init(itemSlice) has less time complexity than N*heap.Push(item), but that's not important 😂